### PR TITLE
Transaction DeliverMin field (RIPD-930)

### PR DESCRIPTION
**Pull request by miguelportilla**
Analogous to SendMax, DeliverMin means that if we deliver less than that, refuse the payment. This feature will be necessary with future path finding that is trying to convert all of an asset.

Reviewers: @seelabs @nbougalis @JoelKatz

### CODE REVIEW

**Comment by vinniefalco** [comment]
Suggested:
`Add DeliverMin transaction field (RIPD-930)`

**Comment by seelabs** [comment]
With the new jtx framework it's pretty easy to write tests. I'd like some simple transactions that exercise this.

**Comment by seelabs** [comment]
This is still on 0.28.0-b4 and needs to be rebased.

**Comment by vinniefalco** [comment]
ouch 0.28.0-b4?!

**Comment by miguelportilla** [comment]
Addressed feedback, rebased and added unit test.

File: src/ripple/app/tx/impl/Payment.cpp [hunk 1]
```diff
@@ -160,6 +161,53 @@ class Payment
...
+            }
+            if (deliverMin > saDstAmount)
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: deliver min greater than dst: " <<
+                        maxSourceAmount.getFullText();
+
+                return temBAD_AMOUNT;
+            }
+            if (bMax && deliverMin > maxSourceAmount)
```
Comment (seelabs) [inline src/ripple/app/tx/impl/Payment.cpp]: This doesn't look right. These may be different currencies, right? Does it make sense to check `maxSourceAmount`?

Comment (JoelKatz) [inline src/ripple/app/tx/impl/Payment.cpp]: During preCheck, we need to make sure that, if a DeliverMin is specified, it's valid and for the correct asset. (The same currency/issuer as Amount.)

File: src/ripple/app/tx/impl/Payment.cpp [hunk 2]
```diff
@@ -310,7 +359,18 @@ class Payment
                     // TODO: is this right?  If the amount is the correct amount, was
                     // the delivered amount previously set?
                     if (rc.result () == tesSUCCESS && rc.actualAmountOut != saDstAmount)
-                        mEngine->view ().setDeliveredAmount (rc.actualAmountOut);
+                    {
+                        if (bDeliverMin && rc.actualAmountOut <
+                            mTxn.getFieldAmount (sfDeliverMin))
+                        {
+                            rc.setResult (tecPATH_PARTIAL);
```
Comment (seelabs) [inline src/ripple/app/tx/impl/Payment.cpp]: I think I suggested `tec_PATH_DRY` for this. `tecPATH_PARTIAL` is much better.

File: src/ripple/app/tx/impl/Payment.cpp (hunk 1 above)
Comment (seelabs) [inline src/ripple/app/tx/impl/Payment.cpp]: Agreed. But `maxSourceAmount` can be a different currency than `deliverMin`. Even if they are the same currency, I'd expect `deliverMin > maxSourceAmount` not to be an error case. I may have to spend $1 to deliver a min of $1.01.

File: src/ripple/app/tx/impl/Payment.cpp [hunk 3]
```diff
@@ -160,6 +161,53 @@ class Payment
...
+                        maxSourceAmount.getFullText ();
+
+                return temBAD_AMOUNT;
+            }
+            if (deliverMin.getIssuer () != saDstAmount.getIssuer ())
+            {
+                if (m_journal.trace) m_journal.trace <<
+                    "Malformed transaction: deliver min issuer"
+                    "differs from dst: " <<
+                        maxSourceAmount.getFullText ();
```
Comment (seelabs) [inline src/ripple/app/tx/impl/Payment.cpp]: This looks like a copy/paste set of errors with `maxSourceAmount`. I think you mean `saDstAmount` here. (Similar error when writing to the journal in this section).

File: src/ripple/app/tx/tests/DeliverMin.test.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        env.require(balance(bob, XRP(10000)));
+
+        env.trust(USD(600), bob);
+        env.require(owners("bob", 1), lines("bob", 1));
+
+        env(pay(alice, bob, USD(10)), delivermin(XRP(5)), ter(temBAD_AMOUNT));
+        env(pay(alice, bob, USD(10)), delivermin(USD(-5)), ter(temBAD_AMOUNT));
+        env(pay(alice, bob, USD(10)), delivermin(USD(15)), ter(temBAD_AMOUNT));
+        env(pay(alice, bob, USD(10)), sendmax(USD(9)), delivermin(USD(10)),
+            ter(temBAD_AMOUNT));
```
Comment (seelabs) [inline src/ripple/app/tx/tests/DeliverMin.test.cpp]: Thanks for writing the tests - much appreciated. All these tests check against constraints in payment `preCheck`. We need at least one test where `delivermin` passes `preCheck`, but fails the payment tries to deliver an amount less than `delivermin` (maybe set up an offer that doesn't have enough liquidity available).

File: src/ripple/app/tx/tests/DeliverMin.test.cpp [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        testcase("payment deliver minimum");
+
+        Env env(*this);
+        Account const alice("alice");
+        auto const USD = alice["USD"];
+        Account const bob("bob");
+
+        env.fund(XRP(10000), alice, bob);
+        env.require(balance(alice, XRP(10000)));
+        env.require(balance(bob, XRP(10000)));
```
Comment (vinniefalco) [inline src/ripple/app/tx/tests/DeliverMin.test.cpp]: These `require` statements are not needed. First of all, we're not testing the `Env::fund` function and second, `Env::fund` already performs such consistency checks internally:
https://github.com/ripple/rippled/blob/develop/src/ripple/test/jtx/impl/Env.cpp#L171

File: src/ripple/test/jtx/impl/amount.cpp [hunk 6]
```diff
@@ -52,6 +52,21 @@ PrettyAmount::operator AnyAmount() const
     return { amount_ };
 }
 
+template <typename T>
+std::string
+to_places(const T d, std::uint32_t places)
```
Comment (vinniefalco) [inline src/ripple/test/jtx/impl/amount.cpp]: Please make this `static` and split it to its own commit

Comment (vinniefalco) [inline src/ripple/test/jtx/impl/amount.cpp]: Should be `static`

Comment (miguelportilla) [inline src/ripple/test/jtx/impl/amount.cpp]: Well 8 bits is is quite optimistic too. I'll add an assert on places though.

File: src/ripple/app/tx/impl/Payment.cpp (hunk 3 above)
Comment (miguelportilla) [inline src/ripple/app/tx/impl/Payment.cpp]: Definitely copy/paste error. I meant deliverMin.

File: src/ripple/app/tx/impl/Payment.cpp (hunk 1 above)
Comment (miguelportilla) [inline src/ripple/app/tx/impl/Payment.cpp]: @seelabs Great catch! Thanks.

File: src/ripple/app/tx/tests/DeliverMin.test.cpp (hunk 5 above)
Comment (miguelportilla) [inline src/ripple/app/tx/tests/DeliverMin.test.cpp]: You may want to update Env_test.cpp:
https://github.com/ripple/rippled/blob/develop/src/ripple/test/jtx/impl/Env_test.cpp#L295-L298

Comment (vinniefalco) [inline src/ripple/app/tx/tests/DeliverMin.test.cpp]: Why? Unlike the DeliverMin test, the `Env_test` is specifically designed to test the jtx framework!

File: src/ripple/app/tx/impl/Payment.cpp [hunk 7]
```diff
@@ -160,6 +161,55 @@ class Payment
                 "No ripple direct specified for XRP to XRP.";
             return temBAD_SEND_XRP_NO_DIRECT;
         }
+        if (bDeliverMin)
+        {
+            auto const deliverMin = mTxn.getFieldAmount (sfDeliverMin);
```
Comment (seelabs) [inline src/ripple/app/tx/impl/Payment.cpp]: Vinnie made a comment on slack that asked about partial payment. I think we need to check that partial payment is set or deliver min doesn't make sense.

Comment (miguelportilla) [inline src/ripple/app/tx/impl/Payment.cpp]: He asked as we were discussing it. Thanks for mentioning it here though.

**Comment by miguelportilla** [comment]
Feedback addressed.

**Comment by seelabs** [comment]
:+1:

**Comment by MarkusTeufelberger** [comment]
Does this need a "Tx Change" label?

**Comment by vinniefalco** [comment]
Yes it do

**Comment by nbougalis** [comment]
Left a minor comment, but it's fine as is. :+1:

**Comment by miguelportilla** [comment]
Addressed Nik's comment, thanks Nik. Will rebase after all tests clear.

**Comment by seelabs** [comment]
@miguelportilla With all the rebasing going on today I wanted to make sure you saw the comments I made on this commit: https://github.com/miguelportilla/rippled/commit/07ca8db830aee09d20f4e10989dd7a4126a91124?diff=unified
